### PR TITLE
Fine-tuning Model Parallel Megatron: Load checkpoint and start training

### PR DESCRIPTION
**Pull request by ericharper**
In order to use model parallel for fine-tuning in NeMo, the user only has to pass the path to the model parallel checkpoints. Since model parallel checkpoints require a specific directory structure, NeMo will infer the model parallel size before instantiating the Megatron BERT encoder.

Note this PR does NOT guarantee convergence, that will come in future PRs when we add model parallel gradient clipping, sampling, etc.

# Usage
This should work for any NLP example script that uses a BERT encoder.
```
#!/bin/bash
export NEMO_TESTING=true

python token_classification.py \
    trainer.gpus=[2,3] \
    trainer.num_nodes=1 \
    model.dataset.data_dir=/raid/data/token_classification/data \
    model.train_ds.batch_size=4 \
    model.language_model.pretrained_model_name=megatron-bert-uncased \
    model.language_model.lm_checkpoint=/raid/results/1194279-2/iter_2000000 \
    model.language_model.config_file=/raid/megatron_checkpoints/mp_bert_345M/config.json 
```
# Changes to NeMo, implementation details
The crucial thing to understand is that Megatron-LM requires torch.distributed to be initialized in order to instantiate Megatron BERT encoders. This poses a challenge as PyTorch Lightning only initializes torch.distributed after calling trainer.fit().

@borisfom figured out that we can lazily initialize megatron-lm, and only fully instantiate the megatron-lm bert encoder after pytorch lightning has initialized torch.distributed by finishing megatron-lm initialization in the first forward pass. 

This is why we lazily load model parallel checkpoints in the first forward pass, because we need the model parallel groups to have been initialized, and to do that we need torch.distributed to be initialized.

Another interesting point, is that in order to instantiate megatron-lm model parallel encoders, we need to know not just the model parallel size but also the world size. See https://github.com/NVIDIA/Megatron-LM/blob/46a536cc888dcd4899c3f74d048ec29b9d93a412/megatron/mpu/initialize.py#L59 .

To get the world size, we need to know num_gpus and num_nodes, note this information is not available from torch.distributed as it hasn't been initialized yet. So we get this information from trainer, and store it in AppState() (app_state.py). This is done in the ModelPT constructor. @tkornuta-nvidia has some ideas on simply extracting that information from the config rather than passing the trainer to ModelPT.

Finally, to correctly pass model_parallel_size to megatron-lm we have to use the ```extra_args_provider``` since we can't pass the arguments with argparser in NeMo (we are using Hydra for configuration).

### CODE REVIEW

**Comment by ericharper** [comment]
closing since CI doesn't see this PR, will create a new one.

**Review by borisfom — APPROVED** [review]

File: nemo/collections/nlp/modules/common/megatron/megatron_utils.py [hunk 1]
```diff
@@ -81,19 +82,24 @@ def get_megatron_lm_model(
...
+            fixed_config = {}
+            for key in config.keys():
+                fixed_key = key.replace("-", "_")
+                if fixed_key == 'max_seq_length':
+                    fixed_key = 'max_position_embeddings'
+                fixed_config[fixed_key] = config[key]
+            # 'vocab_size" no longer used.
+            if 'vocab_size' in fixed_config:
+                fixed_config.pop('vocab_size')
+            config = fixed_config
```
Comment (ekmb) [inline nemo/collections/nlp/modules/common/megatron/megatron_utils.py:102]: Why do we need these changes here? I don't remember if the configs on NGC have `-` or `_`

File: nemo/collections/nlp/modules/common/megatron/megatron_bert.py [hunk 2]
```diff
@@ -56,17 +63,34 @@ def __init__(self, model_name, config, vocab_file):
 
         config['onnx_safe'] = True
 
+        # if 'model_parallel_size' in config:
+        if self._model_parallel_size is not None:
+            app_state = AppState()
+
+            # must be set for model parallel megatron-lm
+            os.environ["WORLD_SIZE"] = str(app_state.world_size)
+
```
Comment (ekmb) [inline nemo/collections/nlp/modules/common/megatron/megatron_bert.py:72]: Setting world size here with the config/trainer doesn't work?

File: nemo/collections/nlp/modules/common/megatron/megatron_utils.py (hunk 1 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/megatron_utils.py:102]: The previous code was hard coding the change so I'm assuming the old configs had the dashes. I'm hoping we can move away from the .json files soon and just input config from CLI or .yaml.

File: nemo/collections/nlp/modules/common/megatron/megatron_bert.py (hunk 2 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/megatron_bert.py:72]: That's right, megatron-lm uses that specific global variable which isn't set when we are instantiating megatron bert in nemo.

File: nemo/collections/nlp/modules/common/megatron/megatron_utils.py (hunk 1 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/megatron/megatron_utils.py:102]: let' not move from .json for megatron as this is want released on ngc

**Review by ekmb — APPROVED** [review]
